src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -170,8 +173,85 @@ StorageObjectStorage::StorageObjectStorage(
         }
     }
 
-    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(
-        metadata.columns, context, sample_path, format_settings, configuration->isDataLakeConfiguration()));
+    /*
+     * If `partition_strategy=hive`, the partition columns shall be extracted from the `PARTITION BY` expression.
+     * There is no need to read from the filepath.
+     *
+     * Otherwise, in case `use_hive_partitioning=1`, we can keep the old behavior of extracting it from the sample path.
+     * And if the schema was inferred (not specified in the table definition), we need to enrich it with the path partition columns
+     */
+    if (configuration->partition_strategy && configuration->partition_strategy_type == PartitionStrategyFactory::StrategyType::HIVE)
+    {
+        hive_partition_columns_to_read_from_file_path = configuration->partition_strategy->getPartitionColumns();
+    }
+    else if (context->getSettingsRef()[Setting::use_hive_partitioning])
+    {
+        hive_partition_columns_to_read_from_file_path = HivePartitioningUtils::extractHivePartitionColumnsFromPath(columns, sample_path, format_settings, context);
```

Comment:
I guess we could add an assertion that all partition columns extracted from the path are present in the storage columns in case these columns were manually specified.

I will not add this assertion, tho. I want to minimize changes and avoid backwards incompatible changes.